components/autocomplete/index.js

```diff
@@ -34,86 +99,194 @@ class Autocomplete extends Component {
-		}
-
-		// Create matching group for the search value
-		pattern += '([^\\s]*)$';
-
-		pattern = new RegExp( pattern );
-
-		let match, search;
-		if ( ( match = value.match( pattern ) ) ) {
-			search = new RegExp( escapeStringRegexp( match[ 1 ] ), 'i' );
-		}
+	// this method is separate so it can be overrided in tests
+	createRange( startNode, startOffset, endNode, endOffset ) {
+		const range = document.createRange();
+		range.setStart( startNode, startOffset );
+		range.setEnd( endNode, endOffset );
+		return range;
+	}
 
-		if ( isOpen ) {
+	search( event ) {
+		const createRange = this.createRange;
+		const { completers: allCompleters } = this.props;
+		const allowAnything = () => true;
+		const { open } = this.state;
+		// ensure that the cursor location is unambiguous
+		const container = event.target;
+		const cursor = this.getCursor( container );
+		// look for the trigger prefix and search query just before the cursor location
+		const match = ( function() {
+			let endTextNode;
+			let endIndex;
+			let completers = allCompleters.map( ( completer, idx ) => ( { ...completer, idx } ) );
+			if ( open !== null ) {
+				// put the open completer at the start so it has priority
+				completers = [
+					open,
+					...completers.filter( ( completer ) => completer.idx !== open.idx ),
+				];
+			}
+			// search backwards to find the first preceeding space or non-text node.
+			if ( isTextNode( cursor.node ) ) { // TEXT node
+				endTextNode = cursor.node;
+				endIndex = cursor.offset;
+			} else if ( cursor.offset === 0 ) {
+				endTextNode = onlyTextNode( descendFirst( cursor.node ) );
+				endIndex = 0;
+			} else {
+				endTextNode = onlyTextNode( descendLast( cursor.node.childNodes[ cursor.offset - 1 ] ) );
+				endIndex = endTextNode ? endTextNode.nodeValue.length : 0;
+			}
+			if ( endTextNode === null ) {
+				return null;
+			}
+			// filter the completers to those that could handle this node
+			completers = completers.filter( ( { allowNode = allowAnything } ) => allowNode( endTextNode, container ) );
+			// exit early if nothing can handle it
+			if ( completers.length === 0 ) {
+				return null;
+			}
+			let startTextNode = endTextNode;
+			let text = endTextNode.nodeValue.substring( 0, endIndex );
+			let pos = lastIndexOfSpace( text );
+			while ( pos === -1 ) {
+				const prev = onlyTextNode( startTextNode.previousSibling );
+				if ( prev ) {
+					// filter the completers to those that could handle this node
+					completers = completers.filter( ( { allowNode = allowAnything } ) => allowNode( endTextNode, container ) );
+					// exit early if nothing can handle it
+					if ( completers.length === 0 ) {
+						return null;
+					}
+					startTextNode = prev;
+					text = prev.nodeValue + text;
+					pos = lastIndexOfSpace( text, prev.nodeValue.length - 1 );
+				} else {
+					break;
+				}
+			}
+			// exit early if nothing can handle it
+			if ( text.substr( pos + 1 ).length === 0 ) {
```

Comment:
Another optimization: We probably don't need to do a full `substr` if we're just testing if there's more text left, instead looking at the immediate next character:

```js
if ( text[ pos + 1 ] === undefined ) {
```

Reply:
Or:
```javascript
if ( text.length <= pos + 1 )
```